### PR TITLE
convert array ops to library calls

### PR DESCRIPTION
**Pull request by MartinNowak**
- capture expression tree as reverse polish notation
- pass scalar subtrees as single argument

Requires <s>https://github.com/dlang/druntime/pull/1891</s>, <s>https://github.com/dlang/druntime/pull/1899</s>, https://github.com/dlang/druntime/pull/1903
Supplemental <s>https://github.com/dlang/dlang.org/pull/1846</s>

### CODE REVIEW

File: test/fail_compilation/fail_arrayop3b.d [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// REQUIRED_ARGS: -o-
+
+/*
+example output from druntime
+----
+../../druntime/import/core/internal/arrayop.d(160): Error: static assert  "Binary op '+=' not supported for element type string."
```
Comment (WalterBright) [inline test/fail_compilation/fail_arrayop3b.d]: Using backticks instead of ' will enable color syntax highlighting!

**Comment by WalterBright** [comment]
Very nice work. Will approve as soon as the changes are incorporated.

File: test/fail_compilation/fail_arrayop3b.d (hunk 1 above)
Comment (MartinNowak) [inline test/fail_compilation/fail_arrayop3b.d]: Surprising that it also works for static asserts, will change.

**Comment by WalterBright** [comment]
@MartinNowak needs a rebase

**Comment by MartinNowak** [comment]
Done @WalterBright

**Comment by ZombineDev** [comment]
I really like the approach taken here. I wonder if this could be used for third-party linear algebra libraries like mir/ndslice. I.e. use the standard function template overloading mechanism - libraries would provide `arrayOp` overloads and the compiler would choose the most specialized one. Of course for array operands, druntime's overload would be the most specialized one.

**Comment by MartinNowak** [comment]
> I really like the approach taken here, but I wonder if this could be useful for third-party linear algebra libraries like mir/ndslice. I.e. use the standard function template overloading mechanism - libraries would provide arrayOp overloads and the compiler would choose the most specialized one. Of course for array operands, druntime's overload would be the most specialized one.

Extensibility was not really a requirement for the story.
A simple extension point would be to change the qualified `object._arrayOp` to an unqualified `_arrayOp`.
https://github.com/dlang/dmd/pull/7032/files#diff-01366d4347de4bb6861fc2c52694b84aR133
Extensibility would also make sense for the normal array operands, e.g. for huge arrays and runtime-CPU/GPU/FPGA specializations.
We don't have much experience with such extension points yet, but if there is interest, it should be trivial for anyone to change the compiler to experiment with the idea.

**Comment by MartinNowak** [comment]
Need https://github.com/dlang/druntime/pull/1899 to fix runnable/arrayop.d.

**Comment by MartinNowak** [comment]
Yikes, someone bluntly added mixed type vector operations (int -> float, even worse int -> double) conversions (https://github.com/dlang/dmd/pull/4218).
I will only support scalar loops for those, as using complex core.simd chains is completely out of scope.
Fortunately gdc/ldc will still be able to auto-vectorize such loops.

**Comment by MartinNowak** [comment]
Done https://github.com/dlang/druntime/pull/1903.

**Comment by MartinNowak** [comment]
Rebased to trigger a retest now that the druntime pulls are merged.

**Pull request merged**